File: src/Blocks.LMT.Client/BlocksLogger.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text.RegularExpressions; // For parsing message templates
using Microsoft.Extensions.Logging; // For LogLevel enum mapping

namespace SeliseBlocks.LMT.Client
{
    public class BlocksLogger : IBlocksLogger
    {
        private readonly LmtOptions _options;
        private readonly ConcurrentQueue<LogData> _logBatch;
        private readonly Timer _flushTimer;
        private readonly LmtServiceBusSender _serviceBusSender;
        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
        private bool _disposed;

        public BlocksLogger(LmtOptions options)
        {
            _options = options ?? throw new ArgumentNullException(nameof(options));

            if (string.IsNullOrWhiteSpace(_options.ServiceId))
                throw new ArgumentException("ServiceName is required", nameof(options));

            if (string.IsNullOrWhiteSpace(_options.ConnectionString))
                throw new ArgumentException("ServiceBusConnectionString is required", nameof(options));

            _logBatch = new ConcurrentQueue<LogData>();

            _serviceBusSender = new LmtServiceBusSender(
                _options.ServiceId,
                _options.ConnectionString,
                _options.MaxRetries,
                _options.MaxFailedBatches);

            var flushInterval = TimeSpan.FromSeconds(_options.FlushIntervalSeconds);
            _flushTimer = new Timer(async _ => await FlushBatchAsync(), null, flushInterval, flushInterval);
        }

        public void Log(LmtLogLevel level, string messageTemplate, Exception? exception = null, params object?[] args)
        {
            if (!_options.EnableLogging) return;

            var activity = Activity.Current;
            var properties = new Dictionary<string, object>();
            string formattedMessage = FormatLogMessage(messageTemplate, args, properties);

            var logData = new LogData
            {
                Timestamp = DateTime.UtcNow,
                Level = level.ToString(),
                Message = formattedMessage,
                Exception = exception?.ToString() ?? string.Empty,
                ServiceName = _options.ServiceId,
                Properties = properties, // Use the parsed properties
                TenantId = _options.XBlocksKey
            };

            if (activity != null)
            {
                logData.Properties["TraceId"] = activity.TraceId.ToString();
                logData.Properties["SpanId"] = activity.SpanId.ToString();
            }

            _logBatch.Enqueue(logData);

            if (_logBatch.Count >= _options.LogBatchSize)
            {
                // Do not await to avoid blocking the caller of Log.
                // This will run on a ThreadPool thread.
                _ = Task.Run(() => FlushBatchAsync());
            }
        }

        public void LogTrace(string messageTemplate, params object?[] args)
            => Log(LmtLogLevel.Trace, messageTemplate, null, args);

        public void LogDebug(string messageTemplate, params object?[] args)
            => Log(LmtLogLevel.Debug, messageTemplate, null, args);

        public void LogInformation(string messageTemplate, params object?[] args)
            => Log(LmtLogLevel.Information, messageTemplate, null, args);

        public void LogWarning(string messageTemplate, params object?[] args)
            => Log(LmtLogLevel.Warning, messageTemplate, null, args);

        public void LogError(string messageTemplate, Exception? exception = null, params object?[] args)
            => Log(LmtLogLevel.Error, messageTemplate, exception, args);

        public void LogCritical(string messageTemplate, Exception? exception = null, params object?[] args)
            => Log(LmtLogLevel.Critical, messageTemplate, exception, args);


        private string FormatLogMessage(string messageTemplate, object?[] args, Dictionary<string, object> properties)
        {
            if (args.Length > 0)
            {
                // Add each arg as Arg0, Arg1, etc. to the properties dictionary
                for (int i = 0; i < args.Length; i++)
                {
                    if (!properties.ContainsKey($"Arg{i}"))
                        properties[$"Arg{i}"] = args[i];
                }


                var formatted = messageTemplate;
                var regex = new Regex(@"\{(.*?)\}");

                int index = 0;
                formatted = regex.Replace(formatted, match =>
                {
                    // If we run out of args, leave placeholder as-is
                    if (index >= args.Length)
                        return match.Value;

                    var value = args[index]?.ToString() ?? string.Empty;
                    index++;
                    return value;
                });

                return formatted;
            }

            return messageTemplate;
        }



        private async Task FlushBatchAsync()
        {
            await _semaphore.WaitAsync();
            try
            {
                var logs = new List<LogData>();
                while (_logBatch.TryDequeue(out var log))
                {
                    logs.Add(log);
                }

                if (logs.Count > 0)
                {
                    await _serviceBusSender.SendLogsAsync(logs);
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"Error flushing logs: {ex}");
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public void Dispose()
        {
            if (_disposed) return;

            _flushTimer?.Change(Timeout.Infinite, Timeout.Infinite);
            _flushTimer?.Dispose();

            _semaphore.Wait();
            try
            {
                FlushBatchAsync().GetAwaiter().GetResult();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"Error during synchronous flush on dispose: {ex}");
            }
            finally
            {
                _semaphore.Release();
                _semaphore?.Dispose();
            }

            _serviceBusSender?.Dispose();

            _disposed = true;
            GC.SuppressFinalize(this);
        }
    }
}

File: src/ruby-lmt-client/lib/selise_blocks/lmt/blocks_logger.rb
# frozen_string_literal: true

require 'singleton'
require 'concurrent'
require 'opentelemetry/api'

module SeliseBlocks
  module LMT
    # Main logger class for Blocks LMT Client
    class BlocksLogger
      include Singleton

      def initialize
        @initialized = false
        @mutex = Mutex.new
      end

      def init(config = nil)
        @mutex.synchronize do
          return if @initialized

          @config = config || SeliseBlocks::LMT.configuration
          @config.validate!

          @log_batch = Concurrent::Array.new
          @batch_mutex = Mutex.new

          @service_bus_sender = ServiceBusSender.new(
            @config.service_id,
            @config.connection_string,
            max_retries: @config.max_retries,
            max_failed_batches: @config.max_failed_batches
          )

          # Start flush timer
          @flush_timer = Concurrent::TimerTask.new(execution_interval: @config.flush_interval_seconds) do
            flush_batch
          end
          @flush_timer.execute

          @initialized = true
        end
      end

      def log(level, message_template, exception: nil, **args)
        init unless @initialized
        return unless @config.enable_logging

        span = OpenTelemetry::Trace.current_span
        properties = {}
        formatted_message = format_log_message(message_template, args, properties)

        log_data = LogData.new
        log_data.timestamp = Time.now.utc
        log_data.level = LogLevel.name(level)
        log_data.message = formatted_message
        log_data.exception = exception ? "#{exception.class}: #{exception.message}\n#{exception.backtrace&.join("\n")}" : ''
        log_data.service_name = @config.service_id
        log_data.properties = properties
        log_data.tenant_id = @config.x_blocks_key

        # Add trace context if available
        if span && span.context.valid?
          log_data.properties['trace_id'] = span.context.hex_trace_id
          log_data.properties['span_id'] = span.context.hex_span_id
        end

        @log_batch << log_data

        # Flush if batch size reached
        flush_batch if @log_batch.size >= @config.log_batch_size
      end

      def trace(message_template, **args)
        log(LogLevel::TRACE, message_template, **args)
      end

      def debug(message_template, **args)
        log(LogLevel::DEBUG, message_template, **args)
      end

      def info(message_template, **args)
        log(LogLevel::INFORMATION, message_template, **args)
      end

      def warn(message_template, **args)
        log(LogLevel::WARNING, message_template, **args)
      end

      def error(message_template, exception: nil, **args)
        log(LogLevel::ERROR, message_template, exception: exception, **args)
      end

      def critical(message_template, exception: nil, **args)
        log(LogLevel::CRITICAL, message_template, exception: exception, **args)
      end

      def flush
        flush_batch
      end

      def shutdown
        return unless @initialized

        @flush_timer&.shutdown
        flush_batch # Final flush
        @service_bus_sender&.close
        @initialized = false
      end

      private

      def format_log_message(message_template, args, properties)
        return message_template if args.empty?

        formatted = message_template.dup
        arg_values = args.values

        # Store all args in properties
        args.each_with_index do |(key, value), index|
          properties["arg#{index}"] = value
        end

        # Replace {placeholder} patterns with values
        index = 0
        formatted.gsub!(/\{[^}]*\}/) do |match|
          if index < arg_values.size
            value = arg_values[index]
            index += 1
            value.to_s
          else
            match
          end
        end

        formatted
      end

      def flush_batch
        @batch_mutex.synchronize do
          return if @log_batch.empty?

          logs = @log_batch.dup
          @log_batch.clear

          # Send in background to avoid blocking
          Thread.new do
            @service_bus_sender.send_logs(logs)
          rescue StandardError => e
            puts "Error flushing logs: #{e.message}"
          end
        end
      end
    end
  end
end

